torch/_higher_order_ops/associative_scan.py

```diff
@@ -0,0 +1,173 @@
+import torch._subclasses.functional_tensor
+
+import torch.utils._pytree as pytree
+
+from torch._C import DispatchKey
+from torch._C._functorch import _add_batch_dim, get_unwrapped, maybe_get_bdim
+from torch._higher_order_ops.utils import (
+    _maybe_run_with_interpreter,
+    _set_compilation_env,
+    autograd_not_implemented,
+    unique_graph_id,
+)
+
+from torch._ops import HigherOrderOperator
+from torch._subclasses.fake_tensor import FakeTensorMode
+from torch.fx.experimental.proxy_tensor import (
+    disable_proxy_modes_tracing,
+    make_fx,
+    ProxyTorchDispatchMode,
+    track_tensor_tree,
+)
+
+aten = torch._ops.ops.aten
+
+
+def associative_scan(
+    input: torch.Tensor,
+    dim: int,
+    combine_fn: Callable[[torch.Tensor, torch.Tensor], torch.Tensor],
+) -> torch.Tensor:
+    r"""
+    Performs an inclusive scan with an associative pointwise combine function.
+
+    .. warning::
+        `torch.associative_scan` is a prototype feature in PyTorch. It currently
+        does not support autograd and you may run into miscompiles.
+        Read more about feature classification at:
+        https://pytorch.org/blog/pytorch-feature-classification-changes/#prototype
+
+    This operator requires runtime code generation and so requires support for
+    ``torch.compile``. Further, only CUDA device codegen is supported at the moment.
+
+    Args:
+        input (torch.Tensor): The input tensor
+        dim (int): the dimension to scan over
+        combine_fn (Callable): A binary callable with type (Tensor, Tensor) -> Tensor,
+            which is pure, pointwise, and satisfies the associative property.
+            i.e. ``combine_fn(a, combine_fn(b, c)) == combine_fn(combine_fn(a, b), c)``
+
+    Example::
+
+        def add(x: torch.Tensor, y: torch.Tensor):
+            return x + y
+
+        cumsum = associative_scan(x, dim, combine_fn=add)
+
+    """
+
+    torch._check(isinstance(input, torch.Tensor), "input must be a Tensor")
+    dim = utils.canonicalize_dim(input.ndim, dim)
+    torch._check(callable(combine_fn), "combine_fn must be a callable")
+
+    if torch._dynamo.is_compiling():
+        return associative_scan_op(input, dim, combine_fn)
+
+    if not torch._dynamo.is_dynamo_supported():
+        raise RuntimeError("associative_scan requires dynamo support.")
+
+    with _set_compilation_env(), torch._dynamo.utils.disable_cache_limit():
+        return torch.compile(associative_scan_op, fullgraph=True)(
+            input, dim, combine_fn
+        )
+
+
+associative_scan_op = HigherOrderOperator("associative_scan")
+
+
+def trace_associative_scan(proxy_mode, func_overload, input, dim, combine_fn):
+    pre_dispatch = getattr(proxy_mode, "pre_dispatch", False)
+
+    with disable_proxy_modes_tracing():
+        sample_inputs = (
+            torch.full((), False, dtype=input.dtype, device=input.device),
+            torch.full((), False, dtype=input.dtype, device=input.device),
+        )
```

Comment:
There are two kinds of associative scan, right? One where the function takes in two scalars, and another where the function takes in two tensors. It sounds like this is the former?

Reply:
Yes, this only supports scalar scans.